### DISCUSSION TITLE
Ask Remix: what's the best way to cache JSON responses from 3rd party API?

### DISCUSSION THREAD

**Discussion by emadabdulrahim** [remix-run/remix#1752] [Q&A] [answered]
I'm ~over-engineering~ re-writing my website in remix and using NotionAPI as my CMS. I noticed the latency with NotionAPI is high (usually > 500ms) and I also don't want to potentially get rate limited. SSG is usually the answer for such problems since I'd only hit the CMS API on  build.

## Possible cache layers:
1. Write the JSON responses from notion to disk? Not possible since Vercel doesn't allow cloud functions to write files.
2. Use a DB as cache layer? I've currently implemented an MVP with supabase and it seems to work well, although Supabase isn't that much faster (still hit 400ms-800ms)
3. Use something like Redis? Never used it before
4. Use Cloudflare KV storage? That might work. Values have 64k limitation, which is fine. I haven't used it before though.

### Question
What would Remix folks suggest I do to cache my NotionAPI response so I don't hit their API every time a user visits my site?

Thanks!

**Comment by sergiodxa** [comment, marked as answer]
You could cache the whole loader response with Cache Control in the CDN, or the html for the initial document request.

**Reply by emadabdulrahim** [reply]
That sounds much better.

- How may I specify that given my hosting platform is Vercel?
- How may I invalidate the cache when my Notion content changes?

Thanks!

**Reply by sergiodxa** [reply]
The Cache Control header is independent of the server. It and Http feature. You define it setting the header, and you control the cache time, I recommend you to read about the header or watch Ryan’s videos about caching in the Remix YouTube channel where he had a great explanation on how to get SSG and ISR like behavior with that header.

**Reply by emadabdulrahim** [reply]
Yeah I did watch it https://www.youtube.com/watch?v=3XkU_DXcgl0

I'm a newbie when it comes to Cache-Control headers though and wasn't sure that simply setting `s-maxage` would already make the entire loader response cached on the CDN.

Still not sure what values to set my `s-maxage` vs `state-while-revalidate` though.

One main question is how may I nuke the cache (on-demand) when I edit the content of a post, regardless of the time value set on the cache-control header?

**Reply by emadabdulrahim** [reply]
Found a good read on max-age and swr headers. https://web.dev/stale-while-revalidate/

**Reply by Adam-Collier** [reply]
> One main question is how may I nuke the cache (on-demand) when I edit the content of a post, regardless of the time value set on the cache-control header?

I'd be interested to know this too :) At the moment if we set stale-while-revalidate cache-control headers and update some content, redirecting to the updated page shows the old content until the page is refreshed. Being able to nuke the cache for the updated page would be great!

Additionally is there a way we could use SQLite or LRU cache (or something similar) on the edge for this purpose? That way we have more control over the cache when content is updated

**Reply by ccssmnn** [reply]
@Adam-Collier Manually resetting the cache is a feature that the CDN needs to provide. An example: Vercel resets all cached responses on a deploy [docs](https://vercel.com/docs/concepts/functions/edge-caching#cache-invalidation-and-purging). My Remix site needs less than a Minute to deploy, so triggering a Vercel deploy hook on content updates would do the job here. They are working on APIs for more selective cache invalidation.

**Reply by Adam-Collier** [reply]
@ccssmnn looking at the page you linked they do mention:
 `When the CDN receives a request with Pragma: no-cache (such as when the browser devtools are open), it will revalidate any stale resource synchronously, instead of in the background.`. 
Does this mean we could possibly use `Pragma: no-cache` for this?

**Reply by ccssmnn** [reply]
I'm not quite sure what your goal is. As of now we can:
- Serve fresh content (no cache)
- Serve semi fresh content (stale-while-revalidate)
- Serve old content until we redeploy (cache forever)

If your revalidation period is small, you could just update the content, hit each route yourself and then the following requests will get fresh content. Or when caching forever, you update your content and hit the redeploy-hook, which purges the cache.

`Pragma: no-cache` should only work if the content is stale, if I'm getting it right. Instead of hitting a route and having fresh content on the second request, you will hit the route and have fresh content immediately, which is not really a benefit. 

If your content is not stale, you should still get the "old" data from the cache.

**Reply by Adam-Collier** [reply]
To add some context I have some admin routes where I can add, edit and delete content (which is stored in a database). Let's take a post page as an example, the content is being cached with stale-while-revalidate so the pages stay quick without having to hit the database on every request.

I want to be able to:
* Edit the post content in the admin
* Save the new content to the database
* Redirect to the updated page showing the updated content

As it stands because of the stale-while-revalidate cache-control headers the old content is being served until I refresh the page after redirecting. 

So my thinking here is I either need to:

* Find a way to invalidate the cache 
* Somehow trigger a "rebuild" of the page before the redirect so the updated content is already there

**Reply by ccssmnn** [reply]
A third option could be to make the database fast and close to your server to always serve fresh data. E.g. Cloudflare KV or Fly.io KeyDB / PostgreSQL Clusters
